File: src/main/java/commands/runningchecker/RunningCheckerManager.java
package commands.runningchecker;

import com.google.common.cache.Cache;
import com.google.common.cache.CacheBuilder;
import commands.Command;
import core.DiscordApiManager;
import core.cache.PatreonCache;

import java.time.Duration;
import java.time.Instant;
import java.util.ArrayList;
import java.util.HashMap;

public class RunningCheckerManager {

    private static final RunningCheckerManager ourInstance = new RunningCheckerManager();

    public static RunningCheckerManager getInstance() {
        return ourInstance;
    }

    private RunningCheckerManager() {
    }

    private final Cache<Long, ArrayList<RunningCheckerSlot>> runningCommandsCache = CacheBuilder.newBuilder()
            .expireAfterAccess(Duration.ofMinutes(1))
            .build();

    public synchronized boolean canUserRunCommand(Command command, long userId, int shardId, int maxCalculationTimeSec) {
        ArrayList<RunningCheckerSlot> runningCommandsList = runningCommandsCache.asMap().computeIfAbsent(userId, k -> new ArrayList<>());
        stopAndRemoveOutdatedRunningCommands(runningCommandsList);

        if (runningCommandsList.isEmpty() || runningCommandsList.size() < getMaxAmount(userId)) {
            final RunningCheckerSlot runningCheckerSlot = new RunningCheckerSlot(userId, shardId, maxCalculationTimeSec, command.hasTimeOut());
            runningCommandsList.add(runningCheckerSlot);
            removeOnThreadEnd(command, runningCommandsList, runningCheckerSlot, userId);

            return true;
        }

        return false;
    }

    private void removeOnThreadEnd(Command command, ArrayList<RunningCheckerSlot> runningCommandsList, RunningCheckerSlot runningCheckerSlot, long userId) {
        command.addCompletedListener(() -> {
            synchronized (this) {
                runningCommandsList.remove(runningCheckerSlot);
                if (runningCommandsList.isEmpty())
                    runningCommandsCache.invalidate(userId);
            }
        });
    }

    private int getMaxAmount(long userId) {
        return PatreonCache.getInstance().getUserTier(userId) >= 3 ? 2 : 1;
    }

    private void stopAndRemoveOutdatedRunningCommands(ArrayList<RunningCheckerSlot> runningCommandsList) {
        new ArrayList<>(runningCommandsList).stream()
                .filter(runningCheckerSlot -> Instant.now().isAfter(runningCheckerSlot.getInstant().plusSeconds(runningCheckerSlot.getMaxCalculationTimeSec())))
                .forEach(runningCheckerSlot -> {
                    runningCheckerSlot.stop();
                    runningCommandsList.remove(runningCheckerSlot);
                });
    }

    public synchronized HashMap<Long, ArrayList<RunningCheckerSlot>> getRunningCommandsMap() {
        return new HashMap<>(runningCommandsCache.asMap());
    }

}

File: src/main/java/commands/runningchecker/RunningCheckerManager.kt
package commands.runningchecker

import com.google.common.cache.CacheBuilder
import commands.Command
import core.cache.PatreonCache
import java.time.Duration
import java.time.Instant

object RunningCheckerManager {

    private val runningCommandsCache = CacheBuilder.newBuilder()
        .expireAfterAccess(Duration.ofMinutes(1))
        .build<Long, ArrayList<RunningCheckerSlot>>()

    @JvmStatic
    @get:Synchronized
    val runningCommandsMap: HashMap<Long, ArrayList<RunningCheckerSlot>>
        get() = HashMap(runningCommandsCache.asMap())

    @JvmStatic
    @Synchronized
    fun canUserRunCommand(command: Command, guildId: Long, userId: Long, shardId: Int, maxCalculationTimeSec: Int): Boolean {
        val runningCommandsList = runningCommandsCache.asMap().computeIfAbsent(userId) { ArrayList() }
        stopAndRemoveOutdatedRunningCommands(runningCommandsList)
        if (runningCommandsList.isEmpty() || runningCommandsList.size < getMaxAmount(guildId, userId)) {
            val runningCheckerSlot = RunningCheckerSlot(userId, shardId, maxCalculationTimeSec, command.commandProperties.enableCacheWipe)
            runningCommandsList.add(runningCheckerSlot)
            removeOnThreadEnd(command, runningCommandsList, runningCheckerSlot, userId)
            return true
        }
        return false
    }

    private fun removeOnThreadEnd(command: Command, runningCommandsList: ArrayList<RunningCheckerSlot>, runningCheckerSlot: RunningCheckerSlot,
                                  userId: Long
    ) {
        command.addCompletedListener { remove(runningCommandsList, runningCheckerSlot, userId) }
    }

    @Synchronized
    private fun remove(runningCommandsList: ArrayList<RunningCheckerSlot>, runningCheckerSlot: RunningCheckerSlot, userId: Long) {
        runningCommandsList.remove(runningCheckerSlot)
        if (runningCommandsList.isEmpty()) {
            runningCommandsCache.invalidate(userId)
        }
    }

    private fun getMaxAmount(guildId: Long, userId: Long): Int {
        return if (PatreonCache.getInstance().hasPremium(userId, true) || PatreonCache.getInstance().isUnlocked(guildId)) 2 else 1
    }

    private fun stopAndRemoveOutdatedRunningCommands(runningCommandsList: ArrayList<RunningCheckerSlot>) {
        ArrayList(runningCommandsList).stream()
            .filter { runningCheckerSlot: RunningCheckerSlot -> Instant.now().isAfter(runningCheckerSlot.instant.plusSeconds(runningCheckerSlot.maxCalculationTimeSec.toLong())) }
            .forEach { runningCheckerSlot: RunningCheckerSlot ->
                runningCheckerSlot.stop()
                runningCommandsList.remove(runningCheckerSlot)
            }
    }

}